internal/resource/key_test.go

```diff
@@ -0,0 +1,246 @@
+6UzrUhdYggA3jOuNEaiFCjexpaIgtkmvflb4yPqX8rq6wosfVOtAuUfO1BkXAe9I
+PspZWiL5oYcoSFmXrwiSG5ln0zkVCEeiN9H/xNHFeQKBgBfCRd5Hso0iAgiwUS/T
+OSSGmeqEIC8Krk/G0V9iw4i9OQoBTFeFrja/3JGqUzvoXHJW012MiJ5ErY1bK8du
+Z2uSW/FjsT7HM69XuC0ibPNJ+5Cw7iQJ6QIXTjID3dhv4NywmENhJW71nSyg/RPT
+xV73bGApHGnNU8lCGx/9s1dL
+-----END PRIVATE KEY-----`
+)
+
+func TestNewKey(t *testing.T) {
+	k := NewKey()
+	require.NotNil(t, k)
+	require.NotNil(t, k.Key)
+}
+
+func TestKey_ID(t *testing.T) {
+	var k Key
+	id := k.ID()
+	require.Empty(t, id)
+
+	k = NewKey()
+	id = k.ID()
+	require.Empty(t, id)
+}
+
+func TestKey_Type(t *testing.T) {
+	require.Equal(t, TypeKey, NewKey().Type())
+}
+
+func TestKey_ProcessDefaults(t *testing.T) {
+	k := NewKey()
+	err := k.ProcessDefaults(context.Background())
+	require.NoError(t, err)
+	require.True(t, validUUID(k.ID()))
+
+	k.Key.Id = ""
+	k.Key.CreatedAt = 0
+	k.Key.UpdatedAt = 0
+	require.Equal(t, k.Resource(), &v1.Key{})
+}
+
+func goodKey() Key {
+	k := NewKey()
+	k.ProcessDefaults(context.Background())
+	return k
+}
+
+func TestKey_Validate(t *testing.T) {
+	tests := []struct {
+		name    string
+		Key     func() Key
+		wantErr bool
+		Errs    []*v1.ErrorDetail
+	}{
+		{
+			name:    "empty key isn't valid",
+			Key:     NewKey,
+			wantErr: true,
+			Errs: []*v1.ErrorDetail{
+				{
+					Type: v1.ErrorType_ERROR_TYPE_ENTITY,
+					Messages: []string{
+						"missing properties: 'id'",
+						"Keys must be defined either in JWK or PEM format",
+					},
+				},
+			},
+		},
+		{
+			name:    "need some content",
+			Key:     goodKey,
+			wantErr: true,
+			Errs: []*v1.ErrorDetail{
+				{
+					Type: v1.ErrorType_ERROR_TYPE_ENTITY,
+					Messages: []string{
+						"Keys must be defined either in JWK or PEM format",
+					},
+				},
+			},
+		},
+		{
+			name: "key with jwk and without kid isn't valid",
+			Key: func() Key {
+				k := goodKey()
+				k.Key.Jwk = &v1.JwkKey{}
+				return k
+			},
+			wantErr: true,
+			Errs: []*v1.ErrorDetail{
+				{
+					Type:  v1.ErrorType_ERROR_TYPE_FIELD,
+					Field: "jwk",
+					Messages: []string{
+						"missing properties: 'kid'",
+					},
+				},
+			},
+		},
+		{
+			name: "key with valid jwk is valid",
+			Key: func() Key {
+				k := NewKey()
+				k.Key.Jwk = &v1.JwkKey{}
+				k.ProcessDefaults(context.Background())
+				return k
+			},
+		},
+		{
+			name: "key with invalid pem is not valid",
+			Key: func() Key {
+				k := NewKey()
+				k.Key.Pem = &v1.PemKey{
+					PublicKey:  "1234",
+					PrivateKey: "hunter2",
+				}
+				k.ProcessDefaults(context.Background())
+				return k
+			},
+			wantErr: true,
+			Errs: []*v1.ErrorDetail{
+				{
+					Type:     v1.ErrorType_ERROR_TYPE_FIELD,
+					Field:    "pem.private_key",
+					Messages: []string{"'hunter2' is not valid 'pem-encoded-private-key'"},
+				},
+				{
+					Type:     v1.ErrorType_ERROR_TYPE_FIELD,
+					Field:    "pem.public_key",
+					Messages: []string{"'1234' is not valid 'pem-encoded-public-key'"},
```

Comment:
Something I just thought of, should we modify `isPEMEncodedCertificate()` to ensure it's not expired? We don't have to do it in this PR, but realized we're accepting expired certificates.